In [25]:
import pandas as pd

# 데이터 불러오기

In [26]:
df = pd.read_csv('final.csv', encoding='cp949')

In [27]:
df.head()

사업자등록번호      결산년월       유동자산      매출채권       비유동자산        유형자산  \
0  1018100340  20181231  2667731.0  127320.0  32346444.0   1179096.0   
1  1018100340  20191231  2168837.0   94085.0  34151633.0   1209369.0   
2  1018100340  20201231  2272260.0   78612.0  38329680.0   1091153.0   
3  1018100340  20211231  1175030.0  273522.0  37135811.0    993025.0   
4  1018100772  20181231   422146.0   17011.0  20370096.0  20352846.0   

         자산총계        유동부채       비유동부채  부  채  총  계       자본금  이익잉여금(결손금）  \
0  35014174.0    422872.0  16176067.0  16598939.0  721100.0   3859072.0   
1  36320470.0  10553878.0   5240445.0  15794323.0  721100.0   3708092.0   
2  40601940.0    863144.0  15496117.0  16359261.0  721100.0   4230449.0   
3  38310842.0   9494272.0   4901383.0  14395654.0  721100.0   5169522.0   
4  20792242.0   2684934.0  11404637.0  14089571.0   52000.0  -7410129.0   

         자본총계        매출액    판매비와관리비  영업이익（손실）  법인세비용차감전순손익     법인세비용  \
0  18415235.0  6629483.0  6210864.0  418619.0    3144847.0  635665.0   
1  20526147.0  6541583.0  5698511.0  843072.0    3446082.0  712661.0   
2  24242679.0  5933317.0  5865171.0   68146.0    4180173.0  773416.0   
3  23915187.0  6158586.0  5771585.0  387000.0    4731471.0  907999.0   
4   6702671.0  1127092.0   446655.0  155690.0      20159.0       NaN   

   당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
0  2509182.0      6.161       4.877      236.820     100.000        6.315   
1  2733421.0      7.600       4.300      215.900     100.000       12.900   
2  3406757.0      8.858       5.121      288.321     100.000        1.149   
3  3823473.0      9.690       7.878      418.549     100.000        6.284   
4    20159.0      1.231     -32.209   -12878.763      53.442       13.813   

   매출액순이익률(%)  수지비율(%)       경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
0      37.849   66.562       61.076   93.685  1.000000e+12         0.000   
1      41.800   62.600          NaN   87.100           NaN         0.000   
2      57.417   58.602 -1336705.033   98.852  1.000000e+12         0.000   
3      62.084   55.228  -634411.186   93.716  1.000000e+12         0.000   
4       1.789   98.361       98.496   86.186  2.059900e+01         1.682   

   금융비용대총비용비율(%  부채비율(%)     차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
0  1.000000e+12   90.137  1.000000e+12     52.594       6.411      2.296   
1           NaN   76.900  0.000000e+00     56.500     -23.100     51.400   
2  1.000000e+12   67.481  1.000000e+12     59.708       3.471      3.560   
3  1.000000e+12   60.195  1.000000e+12     62.424     -21.715     39.700   
4  1.918600e+01  210.208  3.834900e+01     32.236     -10.883     40.058   

   비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)    재고자산회전율(회)  매출채권회전율(회)  \
0      87.841       250.381      0.163  1.000000e+12      44.062   
1      25.500       241.400      0.184  1.000000e+12      59.092   
2      63.921       275.719      0.154  1.000000e+12      68.714   
3      20.495       233.749      0.156  1.000000e+12      34.979   
4     170.151      1250.082      0.055  1.000000e+12      69.711   

     매입채무회전율(회)       미수금      매출원가  무형자산  재고자산 기업접두명      기업명 기업접미명  \
0  1.000000e+12       NaN       NaN   NaN   NaN   NaN     대일건설   (주)   
1           NaN       NaN       NaN   NaN   NaN   NaN     대일건설   (주)   
2  1.000000e+12       NaN       NaN   NaN   NaN   NaN     대일건설   (주)   
3  1.000000e+12    3243.0       NaN   NaN   NaN   NaN     대일건설   (주)   
4  1.000000e+12  386880.0  524746.0   NaN   NaN   (주)  쎈츄럴관광호텔   NaN   

                            기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분 국외투자법인여부  \
0     Daeil Construction Co.,Ltd.  제조  중소기업  일반기업     법인     본점        N   
1     Daeil Construction Co.,Ltd.  제조  중소기업  일반기업     법인     본점        N   
2     Daeil Construction Co.,Ltd.  제조  중소기업  일반기업     법인     본점        N   
3     Daeil Construction Co.,Ltd.  제조  중소기업  일반기업     법인     본점        N   
4  Central Tourist Hotel Co.,Ltd.  제조  중소기업  일반기업     법인     본점        N   

  벤처기업여부 상장코드

# 데이터 전처리

In [28]:
df.columns

Index(['사업자등록번호', '결산년월', '유동자산', '매출채권', '비유동자산', '유형자산', '자산총계', '유동부채',
       '비유동부채', '부  채  총  계', '자본금', '이익잉여금(결손금）', '자본총계', '매출액', '판매비와관리비',
       '영업이익（손실）', '법인세비용차감전순손익', '법인세비용', '당기순이익(손실)', '기업순이익률(%)',
       '유보액/총자산(%)', '유보액/납입자본(%)', '매출액총이익률(%)', '매출액영업이익률(%)', '매출액순이익률(%)',
       '수지비율(%)', '경상수지비율', '영업비율(%)', '금융비용대매출액비율(%', '금융비용대부채비율(%)',
       '금융비용대총비용비율(%', '부채비율(%)', '차입금의존도(%)', '자기자본비율(%)', '순운전자본비율(%)',
       '유동부채비율(%)', '비유동부채비율(%)', '부채총계대 매출액(%)', '총자본회전율(회)', '재고자산회전율(회)',
       '매출채권회전율(회)', '매입채무회전율(회)', '미수금', '매출원가', '무형자산', '재고자산', '기업접두명',
       '기업명', '기업접미명', '기업영문명', '업종', '기업규모', '공기업구분', '개인법인구분', '본점지점구분',
       '국외투자법인여부', '벤처기업여부', '상장코드', '산업코드차수', '산업코드1', '산업코드2', '산업코드3',
       '공공기관유형', '중견기업보호여부', '본점기업코드', '설립일자', '설립구분', '상장일자', '주요사업내용', '국가명',
       '홈페이지URL', '대표자명', '직원수', '종료일자', '시작일자', '휴폐업상태', '상태발생일자'],
      dtype='object')

- 기업접두명, 기업명, ,기업접미명, 기업영문명, 산업코드차수, 산업코드2, 산업코드3, 홈페이지URL, 대표자명은 필요없으므로 제거한다.
- 산업코드2, 산업코드3을 없애는 이유는 기존의 산업코드1에 해당하는 사업을 유치하다가 비슷한 직종을 추가한 것으로 판단되기 때문이다.
- 나머지도 분석에 필요없는 요인이기 때문에 제거한다.

In [29]:
df_drop_col_1 = df.drop(['기업접두명', '기업명','기업접미명', '기업영문명', '산업코드차수', '산업코드2', '산업코드3', '홈페이지URL', '대표자명'], axis=1)

- 현재 피처가 매우 많으므로 피처 요약표를 생성해서 확인해보자.

In [6]:
def resumetable(df):
    """
    데이터를 한눈에 파악하기 쉽도록 피처요약표를 생성하는 함수입니다.
    ---------------------------------------
    input = 데이터프레임
    retrun  = 피처요약표(데이터프레임)
    ---------------------------------------
    """
    summary = pd.DataFrame(df.dtypes, columns=['데이터 타입'])
    summary['결측값 개수'] = df.isna().sum().values
    summary['고윳값 개수'] = df.nunique().values
    
    return summary

In [7]:
# row 생략 없이 출력
pd.set_option('display.max_rows', None)
# col 생략 없이 출력
pd.set_option('display.max_columns', None)

resume_table = resumetable(df_drop_col_1)
resume_table

데이터 타입  결측값 개수  고윳값 개수
사업자등록번호         int64       0   35709
결산년월            int64       0      54
유동자산          float64    1005  102924
매출채권          float64   18614   84378
비유동자산         float64    2042  100547
유형자산          float64    7396   94320
자산총계          float64     995  103165
유동부채          float64    1152  102696
비유동부채         float64   11422   83283
부  채  총  계    float64    1077  102941
자본금           float64    1035   10569
이익잉여금(결손금）    float64    1279  102722
자본총계          float64    1003  102999
매출액           float64    5679   97868
판매비와관리비       float64    1381  101915
영업이익（손실）      float64    1102  101943
법인세비용차감전순손익   float64    1286  101788
법인세비용         float64   36476   63180
당기순이익(손실)     float64    1055  101714
기업순이익률(%)     float64      33   33959
유보액/총자산(%)    float64     127   65538
유보액/납입자본(%)   float64     133   99861
매출액총이익률(%)    float64     691   44971
매출액영업이익률(%)   float64     690   44566
매출액순이익률(%)    float64     692   43021
수지비율(%)       float64     164   47996
경상수지비율        float64    5399   65566
영업비율(%)       float64     690   44530
금융비용대매출액비율(%  float64    2200   21074
금융비용대부채비율(%)  float64      94    7743
금융비용대총비용비율(%  float64    1780   21326
부채비율(%)       float64     121   81131
차입금의존도(%)     float64     386   57048
자기자본비율(%)     float64     119   51140
순운전자본비율(%)    float64     119   68305
유동부채비율(%)     float64     132   68620
비유동부채비율(%)    float64    1013   53888
부채총계대 매출액(%)  float64     655   75834
총자본회전율(회)     float64      15    6992
재고자산회전율(회)    float64      14   36516
매출채권회전율(회)    float64    2132   31313
매입채무회전율(회)    float64    3475   40434
미수금           float64   28040   55783
매출원가          float64   18358   86284
무형자산          float64   55046   38778
재고자산          float64   30011   73327
업종             object       0       8
기업규모           object       0       1
공기업구분          object       0       2
개인법인구분         object       0       1
본점지점구분         object       0       2
국외투자법인여부       object       0       2
벤처기업여부         object       0       2
상장코드           object  110462     134
산업코드1         float64     190    1411
공공기관유형         object       0       3
중견기업보호여부       object       0       2
본점기업코드        float64  105220    2468
설립일자          float64    2002    9899
설립구분           object       0       8
상장일자          float64  110161     190
주요사업내용         object     411   24118
국가명            object  104262      64
직원수           float64    1663     759
종료일자          float64   97718     350
시작일자          float64   97718     769
휴폐업상태          object   97718       4
상태발생일자        float64   99718    1102

- 고윳값이 3개 이하인 피처들의 고윳값 확인

In [8]:
eigenvalue_under_three = resume_table.loc[resume_table['고윳값 개수'] <= 3].index

for i in eigenvalue_under_three:
    print(f'{i} 피처의 고윳값은 {df_drop_col_1[i].unique()} 입니다.')

기업규모 피처의 고윳값은 ['중소기업'] 입니다.
공기업구분 피처의 고윳값은 ['일반기업' '공기업'] 입니다.
개인법인구분 피처의 고윳값은 ['법인'] 입니다.
본점지점구분 피처의 고윳값은 ['본점' '지점'] 입니다.
국외투자법인여부 피처의 고윳값은 ['N' 'Y'] 입니다.
벤처기업여부 피처의 고윳값은 ['N' 'Y'] 입니다.
공공기관유형 피처의 고윳값은 ['해당사항없음' '지자체공단/공사' '공기업/준정부기관'] 입니다.
중견기업보호여부 피처의 고윳값은 ['N' 'Y'] 입니다.


- 고윳값이 하나인 피처들은 분석과 예측력 향상에 도움이 되지 않으므로 제거하도록 하자.
    - 기업규모, 개인법인구분
- 고윳값이 2개인 피처들은 추후에 0과 1로 인코딩 해주자.
    - 공기업구분, 본점지점구분, 국외투자법인여부, 벤처기업여부, 중견기업보호여부
- 고윳값이 3개인 공공기관유형 피처는 추후에 원-핫 인코딩을 적용시켜주자.

In [30]:
# 고윳값이 하나인 피처 제거
df_drop_col_2 = df_drop_col_1.drop(['기업규모', '개인법인구분'], axis=1)

- 고윳값이 많은 피처는 상장코드, 산업코드1, 본점기업코드, 주요사업내용, 국가명이다.
    - 상장코드: 해당 기업의 상장 종목을 나타내는 기호 또는 문자
    - 산업코드1: 세부 산업군을 나타내는 기호 또는 문자
    - 본점기업코드: 여러 지점이 있는 경우 식별하기 위한 코드, 해당 코드는 본점을 식별하기 위한 코드임
    - 주요사업내용: 주매출 내용을 나타냄
    - 국가명: 소재지 국가
- 산업코드1와 국가명을 제외하고는 분석이나 예측력 향상에 도움이 되지 않을 것이라고 판단하여 제거한다.
- 산업코드1은 산업코드 분류 기준에 따라 바꿔주고, 국가명은 결측치가 매우 많으므로 확인해보자.

In [34]:
# 데이터프레임 중간 저장
df_drop_col_2.to_csv('ffinal_drop_feature.csv', encoding='cp949', index=False)

In [12]:
# 산업코드의 결측치 확인
tmp = df.loc[df['산업코드1'].isna()][['사업자등록번호', '기업명', '산업코드1', '산업코드2', '산업코드3', '주요사업내용', '대표자명']]

- 대부분의 산업코드 결측치가 주요사업내용도 같이 없는 경우가 있다.
- 삼포실버드림과 디케이메디칼솔루션은 사업자등록번호가 여러 개이다.
    - nice에 검색해본 결과 다른 업종을 가졌는지 확인이 되지 않는다.
    - 재무데이터 확인이 필요해 보인다.
- 그 다음엔 사업자등록번호를 통해 주요사업내용도 찾고, 산업중분류도 찾아보도록 한다.

### 사업자등록번호가 여러 개인 기업 재무데이터 확인

In [13]:
sampo_df = df.loc[(df['기업명'] == '삼포실버드림')]
for year in sampo_df['결산년월'].unique():
    display(sampo_df.loc[sampo_df['결산년월']==year])

사업자등록번호      결산년월       유동자산      매출채권      비유동자산       유형자산  \
442    1018513125  20181231  3856763.0  620956.0  4801370.0  3895558.0   
15410  1208518548  20181231  3856763.0  620956.0  4801370.0  3895558.0   
56356  2208525693  20181231  3856763.0  620956.0  4801370.0  3895558.0   

            자산총계       유동부채      비유동부채  부  채  총  계       자본금  이익잉여금(결손금）  \
442    8658133.0  1613652.0  3557000.0   5170652.0  800000.0   2687481.0   
15410  8658133.0  1613652.0  3557000.0   5170652.0  800000.0   2687481.0   
56356  8658133.0  1613652.0  3557000.0   5170652.0  800000.0   2687481.0   

            자본총계        매출액    판매비와관리비  영업이익（손실）  법인세비용차감전순손익   법인세비용  \
442    3487481.0  7120135.0  3196965.0  348706.0     190749.0  5574.0   
15410  3487481.0  7120135.0  3196965.0  348706.0     190749.0  5574.0   
56356  3487481.0  7120135.0  3196965.0  348706.0     190749.0  5574.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
442     185176.0      4.262       31.04      335.935      49.798        4.897   
15410   185176.0      4.262       31.04      335.935      49.798        4.897   
56356   185176.0      4.262       31.04      335.935      49.798        4.897   

       매출액순이익률(%)  수지비율(%)   경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
442         2.601   97.332  100.539   95.103         2.465         3.463   
15410       2.601   97.332  100.539   95.103         2.465         3.463   
56356       2.601   97.332  100.539   95.103         2.465         3.463   

       금융비용대총비용비율(%  부채비율(%)  차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
442           2.521  148.263     48.453      40.28      25.908      46.27   
15410         2.521  148.263     48.453      40.28      25.908      46.27   
56356         2.521  148.263     48.453      40.28      25.908      46.27   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
442       101.993         72.62      0.841       2.847       12.57   
15410     101.993         72.62      0.841       2.847       12.57   
56356     101.993         72.62      0.841       2.847       12.57   

       매입채무회전율(회)      미수금       매출원가  무형자산       재고자산 기업접두명     기업명 기업접미명  \
442          9.52  38879.0  3574465.0   NaN  2606936.0   (주)  삼포실버드림   NaN   
15410        9.52  38879.0  3574465.0   NaN  2606936.0   (주)  삼포실버드림   NaN   
56356        9.52  38879.0  3574465.0   NaN  2606936.0   (주)  삼포실버드림   NaN   

                    기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분 국외투자법인여부 벤처기업여부 상장코드  \
442    Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   
15410  Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   
56356  Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   

       산업코드차수  산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부     본점기업코드        설립일자  \
442      10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20001024.0   
15410    10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20010816.0   
56356    10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20070801.0   

      설립구분  상장일자 주요사업내용  국가명 홈페이지URL     대표자명  직원수        종료일자        시작일자  \
442     주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   
15410   주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   
56356   주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   

      휴폐업상태      상태발생일자  
442      폐업  20201231.0  
15410    폐업  20201231.0  
56356    폐업  20201231.0

사업자등록번호      결산년월       유동자산      매출채권      비유동자산       유형자산  \
443    1018513125  20191231  4145796.0  895472.0  4896870.0  3754427.0   
15411  1208518548  20191231  4145796.0  895472.0  4896870.0  3754427.0   
56357  2208525693  20191231  4145796.0  895472.0  4896870.0  3754427.0   

            자산총계       유동부채      비유동부채  부  채  총  계       자본금  이익잉여금(결손금）  \
443    9042666.0  1814306.0  3554000.0   5368306.0  800000.0   2874360.0   
15411  9042666.0  1814306.0  3554000.0   5368306.0  800000.0   2874360.0   
56357  9042666.0  1814306.0  3554000.0   5368306.0  800000.0   2874360.0   

            자본총계        매출액    판매비와관리비  영업이익（손실）  법인세비용차감전순손익    법인세비용  \
443    3674360.0  7121578.0  3240969.0  381878.0     202368.0  15489.0   
15411  3674360.0  7121578.0  3240969.0  381878.0     202368.0  15489.0   
56357  3674360.0  7121578.0  3240969.0  381878.0     202368.0  15489.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
443     186879.0        4.2        31.8        359.3        50.9          5.4   
15411   186879.0        4.2        31.8        359.3        50.9          5.4   
56357   186879.0        4.2        31.8        359.3        50.9          5.4   

       매출액순이익률(%)  수지비율(%)  경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
443           2.6     97.2     NaN     94.6           2.6           3.6   
15411         2.6     97.2     NaN     94.6           2.6           3.6   
56357         2.6     97.2     NaN     94.6           2.6           3.6   

       금융비용대총비용비율(%  부채비율(%)  차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
443             2.7    146.1       46.9       40.6        25.8       49.4   
15411           2.7    146.1       46.9       40.6        25.8       49.4   
56357           2.7    146.1       46.9       40.6        25.8       49.4   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
443          96.7          75.4      0.804        2.72       9.392   
15411        96.7          75.4      0.804        2.72       9.392   
56357        96.7          75.4      0.804        2.72       9.392   

       매입채무회전율(회)      미수금       매출원가  무형자산       재고자산 기업접두명     기업명 기업접미명  \
443        10.976  25393.0  3498731.0   NaN  2629008.0   (주)  삼포실버드림   NaN   
15411      10.976  25393.0  3498731.0   NaN  2629008.0   (주)  삼포실버드림   NaN   
56357      10.976  25393.0  3498731.0   NaN  2629008.0   (주)  삼포실버드림   NaN   

                    기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분 국외투자법인여부 벤처기업여부 상장코드  \
443    Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   
15411  Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   
56357  Sampo Silver Dream  제조  중소기업  일반기업     법인     지점        N      N  NaN   

       산업코드차수  산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부     본점기업코드        설립일자  \
443      10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20001024.0   
15411    10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20010816.0   
56357    10.0    NaN    NaN    NaN  해당사항없음        N  1547317.0  20070801.0   

      설립구분  상장일자 주요사업내용  국가명 홈페이지URL     대표자명  직원수        종료일자        시작일자  \
443     주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   
15411   주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   
56357   주식   NaN    NaN  NaN     NaN  김보옥외 1명  NaN  99991231.0  20201231.0   

      휴폐업상태      상태발생일자  
443      폐업  20201231.0  
15411    폐업  20201231.0  
56357    폐업  20201231.0

- 삼포실버드림은 18, 19년도에 결산을 했다.
- 1018513125, 1208518548, 2208525693의 재무데이터가 모두 같으므로 하나만 선택하도록 하자.

In [14]:
dkmedical_df = df.loc[(df['기업명'] == '디케이메디칼솔루션')]
for year in dkmedical_df['결산년월'].unique():
    display(dkmedical_df.loc[dkmedical_df['결산년월']==year])

사업자등록번호      결산년월        유동자산        매출채권       비유동자산       유형자산  \
18195  1228540328  20181231  44320361.0  13287547.0  11605508.0  7434898.0   
59397  2298111558  20181231  44320361.0  13287547.0  11605508.0  7434898.0   
68607  3148522123  20181231  44320361.0  13287547.0  11605508.0  7434898.0   

             자산총계        유동부채      비유동부채  부  채  총  계        자본금  이익잉여금(결손금）  \
18195  55925869.0  11261536.0  1312348.0  12573883.0  1000000.0  42227843.0   
59397  55925869.0  11261536.0  1312348.0  12573883.0  1000000.0  42227843.0   
68607  55925869.0  11261536.0  1312348.0  12573883.0  1000000.0  42227843.0   

             자본총계         매출액     판매비와관리비   영업이익（손실）  법인세비용차감전순손익     법인세비용  \
18195  43351986.0  46980001.0  15335658.0  1292556.0    1911809.0  372090.0   
59397  43351986.0  46980001.0  15335658.0  1292556.0    1911809.0  372090.0   
68607  43351986.0  46980001.0  15335658.0  1292556.0    1911809.0  372090.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
18195  1539719.0      2.939      75.507     4222.784      35.394        2.751   
59397  1539719.0      2.939      75.507     4222.784      35.394        2.751   
68607  1539719.0      2.939      75.507     4222.784      35.394        2.751   

       매출액순이익률(%)  수지비율(%)  경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
18195       3.277   96.027  91.974   97.249         0.014         0.066   
59397       3.277   96.027  91.060   97.249         0.014         0.066   
68607       3.277   96.027  91.974   97.249         0.014         0.066   

       금융비용대총비용비율(%  부채비율(%)     차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
18195         0.014   29.004  1.000000e+12     77.517      59.112     25.977   
59397         0.014   29.004  1.000000e+12     77.517      59.112     25.977   
68607         0.014   29.004  1.000000e+12     77.517      59.112     25.977   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
18195       3.027        26.764      0.893        8.29       3.892   
59397       3.027        26.764      0.893        8.29       3.892   
68607       3.027        26.764      0.893        8.29       3.892   

       매입채무회전율(회)     미수금        매출원가  무형자산       재고자산 기업접두명        기업명 기업접미명  \
18195       7.718  5936.0  30351788.0   NaN  5926796.0   NaN  디케이메디칼솔루션   (주)   
59397       7.718  5936.0  30351788.0   NaN  5926796.0   NaN  디케이메디칼솔루션   (주)   
68607       7.718  5936.0  30351788.0   NaN  5926796.0   NaN  디케이메디칼솔루션   (주)   

                                기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분 국외투자법인여부  \
18195  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     지점        N   
59397  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     본점        N   
68607  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     지점        N   

      벤처기업여부 상장코드  산업코드차수    산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부     본점기업코드  \
18195      N  NaN    10.0      NaN    NaN    NaN  해당사항없음        N  1327237.0   
59397      N  NaN    10.0  46592.0    NaN    NaN  해당사항없음        N        NaN   
68607      N  NaN    10.0      NaN    NaN    NaN  해당사항없음        N  1327237.0   

             설립일자 설립구분  상장일자                     주요사업내용  국가명         홈페이지URL  \
18195  19860201.0   주식   NaN                        NaN  NaN             NaN   
59397  19860120.0   주식   NaN  초음파 기기, X-ray 장비, CT 장비 외  NaN  www.dkms.co.kr   
68607  19860201.0   주식   NaN                        NaN  NaN             NaN   

           대표자명    직원수        종료일자        시작일자 휴폐업상태      상태발생일자  
18195  이준혁외  1명  104.0  99991231.0  20200731.0    폐업  20200107.0  
59397   이준혁외 1명  113.0         NaN         NaN   NaN         NaN  
68607  이준혁외  1명  104.0  99991231.0  20200731.0    폐업  20200107.0

사업자등록번호      결산년월        유동자산        매출채권       비유동자산       유형자산  \
59398  2298111558  20191231  43537509.0  13859727.0  13546827.0  6924085.0   

             자산총계        유동부채      비유동부채  부  채  총  계        자본금  이익잉여금(결손금）  \
59398  57084337.0  10353735.0  1361201.0  11714937.0  1000000.0  44231251.0   

             자본총계         매출액     판매비와관리비   영업이익（손실）  법인세비용차감전순손익     법인세비용  \
59398  45369400.0  55371203.0  15442935.0  1633930.0    2371861.0  368453.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
59398  2003408.0        3.6        77.5       4423.1        30.8          3.0   

       매출액순이익률(%)  수지비율(%)      경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
59398         3.6     95.8 -19675633.9     97.0           0.0           0.0   

       금융비용대총비용비율(%  부채비율(%)  차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
59398           0.0     25.8        0.0       79.5        58.1       22.8   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
59398         3.0          21.2       0.98      10.088        4.08   

       매입채무회전율(회)      미수금        매출원가  무형자산       재고자산 기업접두명        기업명  \
59398       7.948  22981.0  38294337.0   NaN  5050512.0   NaN  디케이메디칼솔루션   

      기업접미명                           기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분  \
59398   (주)  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     본점   

      국외투자법인여부 벤처기업여부 상장코드  산업코드차수    산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부  \
59398        N      N  NaN    10.0  46592.0    NaN    NaN  해당사항없음        N   

       본점기업코드        설립일자 설립구분  상장일자                     주요사업내용  국가명  \
59398     NaN  19860120.0   주식   NaN  초음파 기기, X-ray 장비, CT 장비 외  NaN   

              홈페이지URL     대표자명    직원수  종료일자  시작일자 휴폐업상태  상태발생일자  
59398  www.dkms.co.kr  이준혁외 1명  113.0   NaN   NaN   NaN     NaN

사업자등록번호      결산년월        유동자산        매출채권       비유동자산       유형자산  \
59399  2298111558  20201231  59932016.0  15884636.0  13303904.0  6380090.0   

             자산총계        유동부채      비유동부채  부  채  총  계        자본금  이익잉여금(결손금）  \
59399  73235920.0  21382417.0  1829481.0  23211898.0  1000000.0  48722524.0   

             자본총계         매출액     판매비와관리비   영업이익（손실）  법인세비용차감전순손익     법인세비용  \
59399  50024022.0  69715869.0  15517329.0  4255394.0    5473688.0  982415.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
59399  4491273.0      6.893      66.528     4872.252      28.362        6.104   

       매출액순이익률(%)  수지비율(%)       경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
59399       6.442   92.346 -27480226.86   93.896         0.001         0.003   

       금융비용대총비용비율(%  부채비율(%)     차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
59399         0.001   46.402  1.000000e+12     68.305      52.638     42.744   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
59399       3.657        33.295       1.07      14.004       4.688   

       매입채무회전율(회)      미수금        매출원가  무형자산       재고자산 기업접두명        기업명  \
59399      10.532  18327.0  49943146.0   NaN  4905922.0   NaN  디케이메디칼솔루션   

      기업접미명                           기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분  \
59399   (주)  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     본점   

      국외투자법인여부 벤처기업여부 상장코드  산업코드차수    산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부  \
59399        N      N  NaN    10.0  46592.0    NaN    NaN  해당사항없음        N   

       본점기업코드        설립일자 설립구분  상장일자                     주요사업내용  국가명  \
59399     NaN  19860120.0   주식   NaN  초음파 기기, X-ray 장비, CT 장비 외  NaN   

              홈페이지URL     대표자명    직원수  종료일자  시작일자 휴폐업상태  상태발생일자  
59399  www.dkms.co.kr  이준혁외 1명  113.0   NaN   NaN   NaN     NaN

사업자등록번호      결산년월        유동자산        매출채권       비유동자산       유형자산  \
59400  2298111558  20211231  57507471.0  16155476.0  15499350.0  7137204.0   

             자산총계        유동부채      비유동부채  부  채  총  계        자본금  이익잉여금(결손금）  \
59400  73006821.0  16903481.0  2029600.0  18933080.0  1000000.0  52647459.0   

             자본총계         매출액     판매비와관리비   영업이익（손실）  법인세비용차감전순손익     법인세비용  \
59400  54073740.0  65049205.0  16596557.0  3935280.0    4885576.0  960640.0   

       당기순이익(손실)  기업순이익률(%)  유보액/총자산(%)  유보액/납입자본(%)  매출액총이익률(%)  매출액영업이익률(%)  \
59400  3924935.0      5.372      72.113     5264.746      31.564         6.05   

       매출액순이익률(%)  수지비율(%)        경상수지비율  영업비율(%)  금융비용대매출액비율(%  금융비용대부채비율(%)  \
59400       6.034   92.639  1.000000e+12    93.95         0.005         0.016   

       금융비용대총비용비율(%  부채비율(%)     차입금의존도(%)  자기자본비율(%)  순운전자본비율(%)  유동부채비율(%)  \
59400         0.006   35.013  1.000000e+12     74.067      55.617      31.26   

       비유동부채비율(%)  부채총계대 매출액(%)  총자본회전율(회)  재고자산회전율(회)  매출채권회전율(회)  \
59400       3.753        29.106       0.89      14.712        4.06   

       매입채무회전율(회)     미수금        매출원가  무형자산       재고자산 기업접두명        기업명 기업접미명  \
59400      10.748  5768.0  44517368.0   NaN  3937070.0   NaN  디케이메디칼솔루션   (주)   

                                기업영문명  업종  기업규모 공기업구분 개인법인구분 본점지점구분 국외투자법인여부  \
59400  DK Medical Solutions Co., Ltd.  제조  중소기업  일반기업     법인     본점        N   

      벤처기업여부 상장코드  산업코드차수    산업코드1  산업코드2  산업코드3  공공기관유형 중견기업보호여부  본점기업코드  \
59400      N  NaN    10.0  46592.0    NaN    NaN  해당사항없음        N     NaN   

             설립일자 설립구분  상장일자                     주요사업내용  국가명         홈페이지URL  \
59400  19860120.0   주식   NaN  초음파 기기, X-ray 장비, CT 장비 외  NaN  www.dkms.co.kr   

          대표자명    직원수  종료일자  시작일자 휴폐업상태  상태발생일자  
59400  이준혁외 1명  113.0   NaN   NaN   NaN     NaN

- 디케이메디칼솔루션은 18, 19, 20, 21년도에 결산을 했다.
- 1228540328, 3148522123은 18년도만 결산을 하고, 2298111558는 18~21년도 모두 결산을 했기 때문에 2298111558를 사용하도록 하자.

### 사업자등록번호를 이용한 중분류 찾기

In [21]:
df.loc[df['산업코드1'].isna()][['사업자등록번호', '기업명', '산업코드1', '주요사업내용', '휴폐업상태', '국가명']]

사업자등록번호                      기업명  산업코드1                주요사업내용  \
442     1018513125                   삼포실버드림    NaN                   NaN   
443     1018513125                   삼포실버드림    NaN                   NaN   
452     1018527779                      위니텍    NaN                   NaN   
3038    1058534234                케이씨크린트래이딩    NaN                   NaN   
5904    1078539365              삼양에프앤비여의도지점    NaN                   NaN   
5905    1078539929               삼양에프앤비영등포점    NaN                   NaN   
5925    1078550147                    블루인더스    NaN                   NaN   
7485    1088521136                     레스빠스    NaN                   NaN   
8932    1108509227                 와이비엘(지점)    NaN                   NaN   
8953    1108525853                임오산업서대문지점    NaN                   NaN   
11067   1148529996                      리메드    NaN                   NaN   
13954   1198537598                    청우에이스    NaN                   NaN   
15410   1208518548                   삼포실버드림    NaN                   NaN   
15411   1208518548                   삼포실버드림    NaN                   NaN   
15443   1208543538                   삼덕회계법인    NaN                   NaN   
15451   1208547121                에이비온바이오메드    NaN                   NaN   
17516   1218532655            엘에프네트웍스송림아울렛점    NaN                   NaN   
17517   1218532655            엘에프네트웍스송림아울렛점    NaN                   NaN   
17530   1218539212       한양자동차정비(주)인천공항자동차정    NaN                   NaN   
18195   1228540328                디케이메디칼솔루션    NaN                   NaN   
18201   1228544834         대지이앤씨산업(주)인천동부지점    NaN                   NaN   
18202   1228544834         대지이앤씨산업(주)인천동부지점    NaN                   NaN   
18210   1228549353               성원디앤에프부평지점    NaN                   NaN   
20684   1248539827                     대경산업    NaN                   NaN   
20762   1248572490                현빈개발(주)화성    NaN                   NaN   
20763   1248572490                현빈개발(주)화성    NaN                   NaN   
23132   1258530061          삼익티에이치케이(주)안성공장    NaN       직선운동시스템및부품외 제조외   
23133   1258530061          삼익티에이치케이(주)안성공장    NaN       직선운동시스템및부품외 제조외   
23140   1258532318                  세종에스엠에스    NaN         한식. 일식. 양식 음식   
23152   1258537161            현민지브이티(주)평택공장    NaN                   NaN   
24233   1268515587                  코단에코크리트    NaN                   NaN   
24295   1268545719                   대흥소프트밀    NaN      제빵발효기,냉동기_x000D_   
24306   1268547740            세종에스엠에스세라지오지점    NaN           한식,양식,일식 음식   
24325   1268560160                성우지반엔지니어링    NaN                   NaN   
25536   1278553420                정선레미콘동북공장    NaN                   NaN   
25542   1278557519              두성에코(주)포천지점    NaN                   NaN   
25543   1278557519              두성에코(주)포천지점    NaN                   NaN   
25552   1278561385                      하늘채    NaN                   NaN   
26543   1288555246                  신신체육이십일    NaN                   NaN   
29198   1308552126             한일물산(주)소사직매장    NaN                   NaN   
29204   1308557553                      한소닉    NaN                   NaN   
31634   1328523215              대가운수(주)경기지점    NaN                   NaN   
33964   1348530726                  화백엔지니어링    NaN                   NaN   
34004   1348548056                  한국파비스제약    NaN                   NaN   
35810   1358525885              세종에스엠에스화산지점    NaN      고급음식점(한,일,양식) 음식   
38067   1378516081                  미래수김포지점    NaN                   NaN   
38091   1378529553           나남에이엘디(주) 인천지점    NaN                   NaN   
38099   1378535282                케이디파워김포지점    NaN                   NaN   
38115   1378541033              순환골재협회오류동지점    NaN                   NaN   
38123   1378545719                  경인환경에너지    NaN                   NaN   
38125   1378546547                 풍진건설김포지점    NaN                   NaN   
38126   1378546547                 풍진건설김포지점    NaN         

In [19]:

# 크롤링에 필요한 라이브러리 불러오기
from selenium import webdriver              
from selenium.webdriver import ActionChains  
from selenium.webdriver.common.by import By  
import time
from selenium.common.exceptions import NoAlertPresentException, NoSuchElementException, TimeoutException, ElementNotInteractableException,NoSuchWindowException, NoSuchFrameException, UnexpectedAlertPresentException

# 산업코드가 결측치인 것들 중에서 사업자등록번호를 가져오기
null_list = df.loc[df['산업코드1'].isna()]['사업자등록번호'].unique().tolist()
name_list = df.loc[df['산업코드1'].isna()]['기업명'].unique().tolist()
result_dict = [{'사업자등록번호':[], '기업명':[], '산업코드1':[], '주요사업내용':[]}, {'사업자등록번호':[], '기업명':[]}]

for company_registration_number, corporation_name in zip(null_list, name_list):
    # 01. 비즈인포 검색창 진입 (sleep=2)
    dr = webdriver.Chrome()   
    dr.set_window_size(1000, 1000)                   
    dr.get('https://www.nicebizinfo.com/cm/CM0100M001GE.nice') 
    time.sleep(2)

    # 02. 검색창에 사업자등록번호를 입력하고 검색버튼 누르기 (sleep=5)
    search_box_biz_info = dr.find_element(By.XPATH, "/html/body/div[2]/div/ul/li[1]/div/form/input")     # 비즈인포 검색창
    search_button_biz_info = dr.find_element(By.XPATH, '/html/body/div[2]/div/ul/li[1]/div/form/button') # 비즈인포 검색 버튼
    time.sleep(2)

    act = ActionChains(dr)
    act.send_keys_to_element(search_box_biz_info, company_registration_number).click(search_button_biz_info).perform() # 사업자등록번호 입력, 버튼 클릭 수행
    time.sleep(5)

    try:
        # 03. 검색결과 클릭하기
        search_result = dr.find_element(By.CSS_SELECTOR, 'body > div.cSection > div > div.cTable.sp3.mb60 > table > tbody > tr.bg > th > span.fz14.fwb.ml10.fErr > a')
        search_result.click()
        time.sleep(5)

        # 03. 검색해서 나온 결과에서 주요사업내용 가져오기 (sleep=3)
        click_result = dr.find_element(By.XPATH, '/html/body/div[3]/div/div[3]/table/tbody/tr[2]/td[2]/div/strong')
        time.sleep(2)

        indurstry_code = re.findall('\(([^)]+)', click_result.text)[1:]   # 산업코드 추출
        business_details = re.sub('\(.*\)|\s-\s.*', '',click_result.text) # 주요사업내용 추출

        result_dict[0]["사업자등록번호"].append(company_registration_number)
        result_dict[0]["기업명"].append(corporation_name) 
        result_dict[0]["산업코드1"].append(indurstry_code)
        result_dict[0]["주요사업내용"].append(business_details)

    except NoSuchElementException as e:
        print(f'{corporation_name} {company_registration_number}의 검색 결과가 없습니다.')
        result_dict[1]["사업자등록번호"].append(company_registration_number)
        result_dict[1]["기업명"].append(corporation_name) 

케이씨크린트래이딩 1058534234의 검색 결과가 없습니다.
블루인더스 1078550147의 검색 결과가 없습니다.
에이비온바이오메드 1208543538의 검색 결과가 없습니다.
엘에프네트웍스송림아울렛점 1208547121의 검색 결과가 없습니다.
대지이앤씨산업(주)인천동부지점 1228540328의 검색 결과가 없습니다.
성원디앤에프부평지점 1228544834의 검색 결과가 없습니다.
대경산업 1228549353의 검색 결과가 없습니다.
코단에코크리트 1258537161의 검색 결과가 없습니다.
세종에스엠에스화산지점 1348548056의 검색 결과가 없습니다.
성지디앤디 1448505224의 검색 결과가 없습니다.
한국교육시스템(주)분당지점코레샵 1808500152의 검색 결과가 없습니다.
세종에스엠에스한국관광용품센타지점 2068515864의 검색 결과가 없습니다.
삼진아트팩지점 2068533241의 검색 결과가 없습니다.
닭익는마을 2158547074의 검색 결과가 없습니다.
대신네트웍스(주)춘천지점 2158547337의 검색 결과가 없습니다.
쎄니트해운대지점 2548500899의 검색 결과가 없습니다.
블레이스홀딩스 2578500637의 검색 결과가 없습니다.
비티캠퍼스 2598500477의 검색 결과가 없습니다.
블루네트웍스중부지사 2968500078의 검색 결과가 없습니다.
신한정밀공업(주)대전공장 3128563152의 검색 결과가 없습니다.
범신사세종지사 3148522123의 검색 결과가 없습니다.
에코로화성지점 3148550537의 검색 결과가 없습니다.
에프티이앤이 3198500823의 검색 결과가 없습니다.


UnexpectedAlertPresentException: Alert Text: 해당 회사는 정보가 부족하여 상세정보를 확인 할 수 없습니다.
Message: unexpected alert open: {Alert text : 해당 회사는 정보가 부족하여 상세정보를 확인 할 수 없습니다.}
  (Session info: chrome=106.0.5249.91)
Stacktrace:
Backtrace:
	Ordinal0 [0x00B0C0A3+2212003]
	Ordinal0 [0x00AA2CC1+1780929]
	Ordinal0 [0x009B465D+804445]
	Ordinal0 [0x00A0F41E+1176606]
	Ordinal0 [0x009FEA36+1108534]
	Ordinal0 [0x009D83C9+951241]
	Ordinal0 [0x009D9396+955286]
	GetHandleVerifier [0x00DB9CE2+2746722]
	GetHandleVerifier [0x00DAA234+2682548]
	GetHandleVerifier [0x00B9B34A+524234]
	GetHandleVerifier [0x00B99B60+518112]
	Ordinal0 [0x00AA9FBC+1810364]
	Ordinal0 [0x00AAEA28+1829416]
	Ordinal0 [0x00AAEB15+1829653]
	Ordinal0 [0x00AB8744+1869636]
	BaseThreadInitThunk [0x7781FA29+25]
	RtlGetAppContainerNamedObjectPath [0x77C17B5E+286]
	RtlGetAppContainerNamedObjectPath [0x77C17B2E+238]


In [22]:
# 산업코드1 변경하기
# 10차수 산업코드 중분류 기준으로 변경

In [ ]:
df_onehot_encoding['산업코드1'].isna().sum()

190

In [24]:
# 국가명 확인해보기
df_onehot_encoding['국가명'].unique()

array([nan, '일본', '네덜란드', '미국', '싱가포르', '몰타', '영국', '파나마', '필리핀', '중국',
       '프랑스', '덴마크', '스페인', '말레이시아', '스위스', '홍콩', '세이셸', '독일', '이탈리아',
       '버뮤다', '캐나다', '케이만군도', '아랍에미레이트', '도미니카연방', '버진아일랜드', '호주', '대만',
       '룩셈부르크', '아일랜드', '기타(구주)', '스웨덴', '우즈베키스탄', '브루나이', '오만', '키프러스',
       '벨기에', '인도', '헝가리', '기타', '러시아', '라부안', '오스트리아', '뉴질랜드', '파키스탄',
       '캄보디아', '루마니아', '터키', '이스라엘', '사모아', '노르웨이', '핀란드', '태국', '방글라데시',
       '그리스', '칠레', '레바논', '포르투칼', '인도네시아', '바베이도스', '슬로베니아', '멕시코',
       '라트비아', '우크라이나', '네팔', '폴란드'], dtype=object)

- 확인 결과 결측치는 대한민국이 소재지인 경우 결측치로 나타난 것으로 확인된다.
- 따라서 결측치는 대한민국으로 대체하도록 한다.

In [31]:
df_onehot_encoding['국가명'] = df_onehot_encoding['국가명'].fillna('대한민국')